exporters/stats/stackdriver/src/main/java/io/opencensus/exporter/stats/stackdriver/StackdriverExportUtils.java

```diff
@@ -363,65 +368,90 @@ static Timestamp convertTimestamp(io.opencensus.common.Timestamp censusTimestamp
-      case GCP_GKE_CONTAINER:
-        GcpGkeContainerMonitoredResource gcpGkeContainerMonitoredResource =
-            (GcpGkeContainerMonitoredResource) autoDetectedResource;
-        builder.putLabels(PROJECT_ID_LABEL_KEY, gcpGkeContainerMonitoredResource.getAccount());
-        builder.putLabels("cluster_name", gcpGkeContainerMonitoredResource.getClusterName());
-        builder.putLabels("container_name", gcpGkeContainerMonitoredResource.getContainerName());
-        builder.putLabels("namespace_name", gcpGkeContainerMonitoredResource.getNamespaceId());
-        builder.putLabels("pod_name", gcpGkeContainerMonitoredResource.getPodId());
-        builder.putLabels("location", gcpGkeContainerMonitoredResource.getZone());
-        return;
-      case AWS_EC2_INSTANCE:
-        AwsEc2InstanceMonitoredResource awsEc2InstanceMonitoredResource =
-            (AwsEc2InstanceMonitoredResource) autoDetectedResource;
-        builder.putLabels("aws_account", awsEc2InstanceMonitoredResource.getAccount());
-        builder.putLabels("instance_id", awsEc2InstanceMonitoredResource.getInstanceId());
-        builder.putLabels("region", "aws:" + awsEc2InstanceMonitoredResource.getRegion());
+    Map<String, String> mappings = null;
+    switch (type) {
+      case ResourceKeyConstants.GCP_GCE_INSTANCE_TYPE:
+        builder.setType(GCP_GCE_INSTANCE);
+        mappings = GCP_RESOURCE_MAPPING;
+        break;
+      case ResourceKeyConstants.K8S_CONTAINER_TYPE:
+        builder.setType(GCP_GKE_CONTAINER);
+        mappings = K8S_RESOURCE_MAPPING;
+        break;
+      case ResourceKeyConstants.AWS_EC2_INSTANCE_TYPE:
+        builder.setType(AWS_EC2_INSTANCE);
+        mappings = AWS_RESOURCE_MAPPING;
+        break;
+      default:
+        builder.setType(GLOBAL);
         return;
     }
+
+    if (mappings != null) {
+      Map<String, String> resLabels = autoDetectedResource.getLabels();
+      for (Map.Entry<String, String> entry : mappings.entrySet()) {
+        if (entry.getValue() != null && resLabels.containsKey(entry.getValue())) {
+          builder.putLabels(entry.getKey(), resLabels.get(entry.getValue()));
+        }
+      }
+      return;
+    }
+
     throw new IllegalArgumentException("Unknown subclass of MonitoredResource.");
   }
 
+  private static Map<String, String> getGcpResourceLabelsMappings() {
+    Map<String, String> resourceLabels = new LinkedHashMap<String, String>();
+    resourceLabels.put("project_id", STACKDRIVER_PROJECT_ID_KEY);
+    resourceLabels.put("instance_id", ResourceKeyConstants.GCP_INSTANCE_ID_KEY);
+    resourceLabels.put("zone", ResourceKeyConstants.GCP_ZONE_KEY);
+    return Collections.unmodifiableMap(resourceLabels);
+  }
+
+  private static Map<String, String> getK8sResourceLabelsMappings() {
+    Map<String, String> resourceLabels = new LinkedHashMap<String, String>();
+    resourceLabels.put("project_id", STACKDRIVER_PROJECT_ID_KEY);
+    resourceLabels.put("location", ResourceKeyConstants.GCP_ZONE_KEY);
```

Comment:
Location is a special key in the Stackdriver. It is not the location of the cluster it is the location where these data should go (very strange). As @fabxc mentioned we talked a bit about this. Maybe file an issue (i know you don't change the logic now) to discuss and resolve this issue.

Reply:
Ok sure. https://cloud.google.com/monitoring/api/resources#tag_k8s_container says "**location:** The physical location of the cluster that contains the container."